File: trunk/data/modules/launch.py
from difficulty import usingDifficulty
import vsrandom
import unit
import ship_upgrades
import VS
import sys

def launch (fgname, faction, type,ai, nr_ships, nr_waves, vec, logo=''):
  diff=usingDifficulty()
  if (not diff or (type.find(".blank")==-1)):
    ret = VS.launch (fgname,type,faction,"unit",ai,nr_ships,nr_waves,VS.SafeEntrancePoint (vec,40),logo)
    return ret
  rsize=0.0
  diffic = VS.GetDifficulty()
  ret=VS.Unit()
  for i in range(nr_ships):
    mynew=VS.launch(fgname,type,faction,"unit",ai,1,nr_waves,VS.SafeEntrancePoint (vec,40),logo)
    if (i==0):
      ret = mynew
      rsize =mynew.rSize ()*1.75
    ship_upgrades.upgradeUnit ( mynew,diffic)
    vec=(vec[0]-rsize,
         vec[1],#-rsize
        vec[2]-rsize)
  return ret

def launch_waves_around_area(fgname,faction,type,ai,nr_ships,nr_waves,r1,r2,pos,logo=''):
  pos=((pos[0]+vsrandom.uniform(r1,r2)*vsrandom.randrange(-1,2,2)),
       (pos[1]+vsrandom.uniform(r1,r2)*vsrandom.randrange(-1,2,2)),
       (pos[2]+vsrandom.uniform(r1,r2)*vsrandom.randrange(-1,2,2)))
  VS.playAnimation ("warp.ani",pos,300.0)
  return launch(fgname,faction,type,ai,nr_ships,nr_waves,pos,logo)

def launch_wave_around_area(fgname,faction,type,ai,nr_ships,r1,r2,pos,logo):
  return launch_waves_around_area (fgname,faction,type,ai,nr_ships,1,r1,r2,pos,logo='')

def launch_around_station(station_name,fgname,faction,type,ai,nr_ships,nr_waves,logo=''):
  station_unit=unit.getUnitByFgID(station_name)
  if(station_unit.isNull()):
    sys.stderr.write("launch.py:launch_around_station did not find unit %s\n" % (station_name))
    return VS.Unit()
  station_pos=station_unit.Position()
  rsize=station_unit.rSize()
  launched =launch_waves_around_area(fgname,faction,type,ai,nr_ships,nr_waves,rsize,rsize*2.0,station_pos,logo)
  return launched

launch_around_unit=launch_around_station

def launch_waves_in_area(fgname,faction,type,ai,nr_ships,nr_waves,radius,pos,logo=''):
  pos=(pos[0]+vsrandom.uniform((-radius)/2,radius/2.0),
       pos[1]+vsrandom.uniform((-radius)/2,radius/2.0),
       pos[2]+vsrandom.uniform((-radius)/2,radius/2.0))
  VS.playAnimation ("warp.ani",pos,300.0)
  un = launch(fgname,faction,type,ai,nr_ships,nr_waves,pos,logo)

def launch_wave_in_area(fgname,faction,type,ai,nr_ships,radius,pos,logo):
  launch_waves_in_area(fgname,faction,type,ai,nr_ships,1,radius,pos,logo)

def launchShipsAtWaypoints(waypoints,faction,type,ainame,nr,logo=''):
  i=0
  for wp in waypoints:
    outstr="wp%d" % (i)
    VS.playAnimation ("warp.ani",wp,300.0)
    launch(outstr,faction,type,ainame,nr,1,wp,logo)
    i+=1

def launch_wave_around_unit (fgname, faction, type, ai, nr_ships, minradius, maxradius, my_unit,logo=''):
  myvec = (0,0,0)
  if (my_unit.isNull()):
    un=launch_wave_around_area (fgname,faction,type,ai,nr_ships,minradius,maxradius,myvec,logo)
    return un
  myvec=my_unit.LocalPosition()
  print myvec
  rsiz=my_unit.rSize()
  un=launch_wave_around_area (fgname,faction,type,ai,nr_ships,rsiz+minradius,rsiz+maxradius,myvec,logo)
  return un

def launch_wave_around_significant (fgname,faction,type,ai,nr_ships,minradius, maxradius,significant_number,logo=''):
  significant_unit=unit.getSignificant(significant_number,0,0)
  if (significant_unit.isNull()):
    significant_unit = VS.getPlayer()
  launched = launch_wave_around_unit(fgname,faction,type,ai,nr_ships,minradius,maxradius,significant_unit,logo)
  return launched

File: trunk/data/modules/launch.c
module launch {

  import random;
  import unit;

  bool launch_around_station(object station_name,object fgname,object faction,object type,object ai,int nr_ships,int nr_waves){
    object station_unit=unit.getUnitByFgID(station_name);
    if(_std.isNull(station_unit)){
      _io.printf("launch.c:launch_around_station did not find unit %s\n",station_name);
      return false;
    }
    object station_pos=_unit.getPosition(station_unit);

    float rsize=_unit.getRSize(station_unit);

    launch_waves_around_area(fgname,faction,type,ai,nr_ships,nr_waves,rsize,rsize*2.0,station_pos);
    _olist.delete(station_pos);

    return true;
  };

  bool launch_around_unit(object station_name,object fgname,object faction,object type,object ai,int nr_ships,int nr_waves){
    bool ok=launch_around_station(station_name,fgname,faction,type,ai,nr_ships,nr_waves);

    return ok;
  };

  object launch_wave_around_area(object fgname,object faction,object type,object ai,int nr_ships,float r1,float r2,object pos){
    

    float x=_olist.at(pos,0)+(random.random(r1,r2)*random.randomsign());
    float y=_olist.at(pos,1)+(random.random(r1,r2)*random.randomsign());
    float z=_olist.at(pos,2)+(random.random(r1,r2)*random.randomsign());

    // _io.printf("launching %d ships type=%s faction=%s around area [%f,%f,%f]\n",nr_ships,type,faction,x,y,z);

    object un=_unit.launch(fgname,faction,type,ai,nr_ships,1,x,y,z);
    return un;
  };
  void launch_waves_around_area(object fgname,object faction,object type,object ai,int nr_ships,int nr_waves,float r1,float r2,object pos){
    

    float x=_olist.at(pos,0)+(random.random(r1,r2)*random.randomsign());
    float y=_olist.at(pos,1)+(random.random(r1,r2)*random.randomsign());
    float z=_olist.at(pos,2)+(random.random(r1,r2)*random.randomsign());

    // _io.printf("launching %d ships type=%s faction=%s around area [%f,%f,%f]\n",nr_ships,type,faction,x,y,z);

    _unit.launch(fgname,faction,type,ai,nr_ships,nr_waves,x,y,z);
  };


  void launch_wave_in_area(object fgname,object faction,object type,object ai,int nr_ships,float radius,object pos){

    float x=_olist.at(pos,0)+random.random((0.0-radius)/2.0,radius/2.0);
    float y=_olist.at(pos,1)+random.random((0.0-radius)/2.0,radius/2.0);
    float z=_olist.at(pos,2)+random.random((0.0-radius)/2.0,radius/2.0);

    //    _io.printf("launching %d ships type=%s faction=%s in area [%f,%f,%f]\n",nr_ships,type,faction,x,y,z);

    _unit.launch(fgname,faction,type,ai,nr_ships,1,x,y,z);
  };

  void launch_waves_in_area(object fgname,object faction,object type,object ai,int nr_ships,int nr_waves,float radius,object pos){

    float x=_olist.at(pos,0)+random.random((0.0-radius)/2.0,radius/2.0);
    float y=_olist.at(pos,1)+random.random((0.0-radius)/2.0,radius/2.0);
    float z=_olist.at(pos,2)+random.random((0.0-radius)/2.0,radius/2.0);

    //    _io.printf("launching %d ships type=%s faction=%s in area [%f,%f,%f]\n",nr_ships,type,faction,x,y,z);

    _unit.launch(fgname,faction,type,ai,nr_ships,nr_waves,x,y,z);
  };

  void launchShipsAtWaypoints(object waypoints,object faction,object type,object ainame,int nr){
    int i=0;
    object outstr=_string.new();

    while(i<_olist.size(waypoints)){
      _io.sprintf(outstr,"wp%d",i);

      object wp=_olist.at(waypoints,i);
      float x=_olist.at(wp,0);
      float y=_olist.at(wp,1);
      float z=_olist.at(wp,2);

      _unit.launch(outstr,faction,type,ainame,nr,1,x,y,z);
      i=i+1;
    }
  };
  object launch_wave_around_significant (object fgname,object faction,object type,object ai,int nr_ships,float radius,int significant_number) {
    object significant_unit=unit.getSignificant(significant_number);
    object myvec;
    if (_std.isNull(significant_unit)) {
      myvec = _olist.new();
      _olist.push_back(myvec.0);
      _olist.push_back(myvec.0);
      _olist.push_back(myvec.0);
      object un=launch_wave_around_area (fgname,faction,type,ai,nr_ships,0,radius,myvec);
      _olist.delete(myvec);
      return un;
    } else {
      myvec=_unit.getPosition(signigicant_unit);
    }
    float rsiz=_unit.getRSize(significant_unit)*2;
    radius = rsiz+radius;
    object un=launch_wave_around_area (fgname,faction,type,ai,nr_ships,rsiz,radius,myvec);
    return un;
 };

}
